// @inworld/runtime/build/graph/dsl/nodes/custom/CustomNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomNode = void 0;
const nodes_1 = require("../../../../graph/nodes");
const AbstractNode_1 = require("../builtin/AbstractNode");
class CustomNode extends AbstractNode_1.AbstractNode {
    constructor(props) {
        var _a;
        super(props);
        this.inputTypes = [];
        this.executionConfig = (_a = props === null || props === void 0 ? void 0 : props.executionConfig) !== null && _a !== void 0 ? _a : {};
    }
    registerCutomNodeType(inputTypes, outputType, processFunction) {
        this.inputTypes = inputTypes;
        this.outputType = outputType;
        return (0, nodes_1.registerCustomNodeType)(this.constructor.name + 'Type', inputTypes, outputType, processFunction);
    }
    toGraphConfigNode() {
        return {
            type: this.typeRef.typeName,
            id: this.id,
            execution_config: {
                type: 'NodeExecutionConfig',
                properties: Object.assign({ report_to_client: this.reportToClient }, this.executionConfig),
            },
        };
    }
}
exports.CustomNode = CustomNode;
