// @inworld/runtime/build/graph/dsl/components/remote_embedder_component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteEmbedderComponent = void 0;
const common_1 = require("../../../common");
const constants_1 = require("../constants");
const abstract_component_1 = require("./abstract_component");
/**
 * Remote embedder component for text embedding services.
 * Provides access to external embedding providers for converting text into
 * vector representations used for similarity matching and retrieval.
 *
 * @example
 * ```typescript
 * const embedderComponent = new RemoteEmbedderComponent({
 *   provider: 'openai',
 *   modelName: 'text-embedding-ada-002',
 * });
 * ```
 */
class RemoteEmbedderComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new RemoteEmbedderComponent instance.
     *
     * @param {Object} props - Configuration for the remote embedder component
     * @param {string} [props.id] - Optional explicit component identifier. Human-readable identifier would be generated if omitted.
     * @param {string} [props.provider] - Embedder provider (e.g., 'openai', 'inworld')
     * @param {string} [props.modelName] - Provider-specific model name.
     */
    constructor(props) {
        var _a, _b;
        super(props);
        this.provider = (_a = props.provider) !== null && _a !== void 0 ? _a : common_1.DEFAULT_PROVIDER;
        this.modelName = (_b = props.modelName) !== null && _b !== void 0 ? _b : constants_1.DEFAULT_EMBEDDER_MODEL_NAME;
    }
    /**
     * Converts the remote embedder component to a graph configuration component.
     *
     * @returns The configured remote embedder component for graph execution
     */
    toGraphConfigComponent() {
        return {
            id: this.id,
            type: 'TextEmbedderInterface',
            creation_config: {
                type: 'RemoteTextEmbedderConfig',
                properties: {
                    provider: this.provider,
                    model_name: this.modelName,
                },
            },
        };
    }
}
exports.RemoteEmbedderComponent = RemoteEmbedderComponent;
