// @inworld/runtime/build/graph/dsl/components/RemoteSTTComponent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteSTTComponent = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const AbstractComponent_1 = require("./AbstractComponent");
/**
 * Remote STT (Speech-to-Text) component for audio transcription services.
 * This component provides access to external speech recognition services
 * for converting audio input into text transcriptions.
 *
 * @example
 * ```typescript
 * const sttComponent = new RemoteSTTComponent({
 *   id: 'my-stt-component',
 *   sttConfig: {
 *     provider: 'whisper',
 *     modelName: 'whisper-1',
 *     language: 'en'
 *   }
 * });
 * ```
 */
class RemoteSTTComponent extends AbstractComponent_1.AbstractComponent {
    /**
     * Creates a new RemoteSTTComponent instance.
     *
     * @param props - Configuration for the remote STT component
     */
    constructor(props) {
        super(props);
        this.sttConfig = props.sttConfig;
    }
    /**
     * Converts the remote STT component to a graph configuration component.
     *
     * @returns The configured remote STT component for graph execution
     */
    toGraphConfigComponent() {
        return {
            id: this.id,
            type: 'STTInterface',
            creation_config: {
                type: 'RemoteSTTConfig',
                properties: {
                    default_config: (0, snakify_1.default)(this.sttConfig),
                },
            },
        };
    }
}
exports.RemoteSTTComponent = RemoteSTTComponent;
