// @inworld/runtime/build/graph/dsl/nodes/builtin/CustomNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomNode = void 0;
const nodes_1 = require("../../../nodes");
const AbstractNode_1 = require("./AbstractNode");
class CustomNode extends AbstractNode_1.AbstractNode {
    constructor(props = {}) {
        var _a;
        super(props);
        this.executionConfig = (_a = props === null || props === void 0 ? void 0 : props.executionConfig) !== null && _a !== void 0 ? _a : {};
        if (!CustomNode.typeReferences[this.constructor.name]) {
            CustomNode.typeReferences[this.constructor.name] = (0, nodes_1.registerCustomNodeType)(this.constructor.name + 'Type', this.process);
        }
    }
    toGraphConfigNode() {
        return {
            type: this.constructor.name + 'Type',
            id: this.id,
            execution_config: {
                type: 'NodeExecutionConfig',
                properties: Object.assign({ report_to_client: this.reportToClient }, this.executionConfig),
            },
        };
    }
}
exports.CustomNode = CustomNode;
CustomNode.typeReferences = {};
