// @inworld/runtime/build/graph/dsl/nodes/builtin/LLMPromptBuilderNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLMPromptBuilderNode = void 0;
const AbstractNode_1 = require("./AbstractNode");
/**
 * LLM prompt builder node that processes JSON input into formatted prompts.
 * This node transforms structured JSON data into text prompts that can be used
 * directly with LLM completion endpoints.
 *
 * @input {JsonData} {@link JsonData} - The data type that LLMPromptBuilderNode accepts as input
 * @output {TextData} {@link TextData} - The data type that LLMPromptBuilderNode accepts as output
 *
 * @example
 * ```typescript
 * const promptBuilderNode = new LLMPromptBuilderNode({
 *   id: 'prompt-builder',
 *   messages: [
 *     { role: 'system', content: 'You are a helpful assistant.' },
 *     { role: 'user', content: '{{user_question}}' }
 *   ],
 *   responseFormat: 'text'
 * });
 * ```
 */
class LLMPromptBuilderNode extends AbstractNode_1.AbstractNode {
    /**
     * Creates a new LLMPromptBuilderNode instance.
     *
     * @param props - Configuration for the LLM prompt builder node
     */
    constructor(props = {}) {
        super(props);
        this.messages = props.messages;
        this.tools = props.tools;
        this.toolChoice = props.toolChoice;
        this.responseFormat = props.responseFormat;
    }
    /**
     * Converts the LLM prompt builder node to a graph configuration node.
     *
     * @returns The configured LLM prompt builder node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'LLMPromptBuilderNode',
            execution_config: {
                type: 'LLMPromptBuilderNodeExecutionConfig',
                properties: {
                    messages: this.messages,
                    tools: this.tools,
                    tool_choice: this.toolChoice,
                    response_format: this.responseFormat,
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.LLMPromptBuilderNode = LLMPromptBuilderNode;
