// @inworld/runtime/build/graph/dsl/nodes/builtin/MCPCallToolNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPCallToolNode = void 0;
const MCPClientComponent_1 = require("../../components/MCPClientComponent");
const AbstractNode_1 = require("./AbstractNode");
/**
 * MCP call tool node that calls multiple tools on an MCP server in parallel.
 * This node processes tool call requests and executes them against the configured
 * MCP server, returning the results of all tool executions.
 *
 * @input {ToolCallRequestData} {@link ToolCallRequestData} - The data type that MCPCallToolNode accepts as input
 * @output {ToolCallResponseData} {@link ToolCallResponseData} - The data type that MCPCallToolNode accepts as output
 *
 * @example
 * ```typescript
 * const mcpCallToolNode = new MCPCallToolNode({
 *   id: 'mcp-call-tool',
 *   mcpComponentId: 'my-mcp-component',
 *   reportToClient: true
 * });
 * ```
 */
class MCPCallToolNode extends AbstractNode_1.AbstractNode {
    /**
     * Creates a new MCPCallToolNode instance.
     *
     * @param props - Configuration for the MCP call tool node
     */
    constructor(props) {
        super(props);
        let mcpComponent;
        if ('mcpComponent' in props) {
            mcpComponent = props.mcpComponent;
        }
        else {
            mcpComponent = new MCPClientComponent_1.MCPClientComponent({
                id: `${this.id}_mcp_component`,
                sessionConfig: props.sessionConfig,
            });
        }
        this.addComponent(mcpComponent);
    }
    /**
     * Converts the MCP call tool node to a graph configuration node.
     *
     * @returns The configured MCP call tool node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'MCPCallToolNode',
            execution_config: {
                type: 'MCPCallToolNodeExecutionConfig',
                properties: {
                    mcp_component_id: this.getFirstComponent().id,
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.MCPCallToolNode = MCPCallToolNode;
