// @inworld/runtime/build/graph/dsl/nodes/builtin/RemoteLLMCompletionNode.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteLLMCompletionNode = void 0;
const RemoteLLMComponent_1 = require("../../components/RemoteLLMComponent");
const constants_1 = require("../../constants");
const snakify_1 = __importDefault(require("../../utils/snakify"));
const AbstractNode_1 = require("./AbstractNode");
/**
 * Remote LLM completion node for text generation.
 * You can either use a pre-configured LLM component that could be reused across multiple nodes
 * or provide the LLM provider and model name and the node will create a new component for you.
 *
 * @input {TextData} {@link TextData} - The data type that LLMCompletionNode accepts as input
 * @output {TextData | TextStreamData} {@link TextData} | {@link TextStreamData} - The data type that LLMCompletionNode outputs
 *
 * @example
 * ```typescript
 * // Using LLM provider configuration
 * const completionNode = new RemoteLLMCompletionNode({
 *   id: 'my-completion-node',
 *   provider: 'openai',
 *   modelName: 'gpt-4o-mini',
 *   stream: true
 * });
 *
 * // Using existing LLM component
 * const completionNodeWithComponent = new RemoteLLMCompletionNode({
 *   id: 'my-completion-node',
 *   llmComponent: existingLLMComponent
 * });
 *
 * // Using default settings
 * const defaultCompletionNode = new RemoteLLMCompletionNode();
 * ```
 */
class RemoteLLMCompletionNode extends AbstractNode_1.AbstractNode {
    /**
     * Creates a new RemoteLLMCompletionNode instance.
     *
     * @param props - Optional configuration for the LLM completion node. Can specify either LLM provider settings
     *                or reference an existing LLM component, but not both. If not provided, uses default settings.
     */
    constructor(props = {}) {
        var _a, _b, _c, _d, _e;
        super(props);
        this.executionConfig = {
            llmComponentId: '',
            textGenerationConfig: Object.assign({}, (0, snakify_1.default)((_a = props.textGenerationConfig) !== null && _a !== void 0 ? _a : {})),
            stream: (_b = props.stream) !== null && _b !== void 0 ? _b : false,
            reportToClient: (_c = props.reportToClient) !== null && _c !== void 0 ? _c : false,
        };
        let llmComponent;
        if ('llmComponent' in props) {
            llmComponent = props.llmComponent;
        }
        else {
            llmComponent = new RemoteLLMComponent_1.RemoteLLMComponent({
                id: `${this.id}_llm_component`,
                provider: (_d = props.provider) !== null && _d !== void 0 ? _d : constants_1.DEFAULT_PROVIDER,
                modelName: (_e = props.modelName) !== null && _e !== void 0 ? _e : constants_1.DEFAULT_LLM_MODEL_NAME,
            });
        }
        this.addComponent(llmComponent);
        this.executionConfig.llmComponentId = llmComponent.id;
    }
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'LLMCompletionNode',
            execution_config: {
                type: 'LLMCompletionNodeExecutionConfig',
                properties: (0, snakify_1.default)(this.executionConfig),
            },
        };
    }
}
exports.RemoteLLMCompletionNode = RemoteLLMCompletionNode;
