// @inworld/runtime/build/graph/dsl/nodes/builtin/TextChunkingNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextChunkingNode = void 0;
const AbstractNode_1 = require("./AbstractNode");
/**
 * Text chunking node for breaking large text into smaller pieces (a stream of token chunks).
 * This node processes streaming text data and converts it into discrete text chunks.
 *
 * @input {TextStreamData} {@link TextStreamData} - The data type that TextChunkingNode accepts as input
 * @output {TextData} {@link TextData} - The data type that TextChunkingNode accepts as output
 *
 * @example
 * ```typescript
 * const chunkingNode = new TextChunkingNode({
 *   id: 'my-chunking-node',
 *   reportToClient: true
 * });
 * ```
 */
class TextChunkingNode extends AbstractNode_1.AbstractNode {
    /**
     * Creates a new TextChunkingNode instance.
     *
     * @param props - Configuration for the text chunking node
     */
    constructor(props = {}) {
        super(props);
    }
    /**
     * Converts the text chunking node to a graph configuration node.
     *
     * @returns The configured text chunking node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'TextChunkingNode',
            execution_config: {
                type: 'NodeExecutionConfig',
                properties: {
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.TextChunkingNode = TextChunkingNode;
