// @inworld/runtime/build/graph/dsl/nodes/builtin/TextAggregatorNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextAggregatorNode = void 0;
const AbstractNode_1 = require("./AbstractNode");
/**
 * Text aggregator node for combining multiple text pieces into a single text output.
 * This node processes various text inputs and combines them into a unified text result.
 *
 * @input {TextData | TextStreamData | LLMChatResponse} {@link TextData} | {@link TextStreamData} | {@link LLMChatResponseData} - The data type that TextAggregatorNode accepts as input
 * @output {TextData} {@link TextData} - The data type that TextAggregatorNode accepts as output
 *
 * @example
 * ```typescript
 * const aggregatorNode = new TextAggregatorNode({
 *   id: 'my-aggregator-node',
 *   reportToClient: true
 * });
 * ```
 */
class TextAggregatorNode extends AbstractNode_1.AbstractNode {
    /**
     * Creates a new TextAggregatorNode instance.
     *
     * @param props - Configuration for the text aggregator node
     */
    constructor(props = {}) {
        super(props);
    }
    /**
     * Converts the text aggregator node to a graph configuration node.
     *
     * @returns The configured text aggregator node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'TextAggregatorNode',
            execution_config: {
                type: 'NodeExecutionConfig',
                properties: {
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.TextAggregatorNode = TextAggregatorNode;
