// @inworld/runtime/build/core/text_processing/stream_text_processor.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _StreamTextProcessor_destroy;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamTextProcessor = void 0;
const common_1 = require("../../common");
const helpers_1 = require("../../common/helpers");
const stream_1 = require("../../common/stream");
/**
 * Abstract base class for text processing stream operations.
 * Provides common functionality for different text processors.
 * @abstract
 * @internal
 */
class StreamTextProcessor {
    /**
     * Creates a new StreamTextProcessor instance.
     *
     * @param textStream - External reference to the text stream
     * @param destroy - Function to call when destroying the processor
     */
    constructor(textStream, destroy) {
        _StreamTextProcessor_destroy.set(this, void 0);
        __classPrivateFieldSet(this, _StreamTextProcessor_destroy, destroy, "f");
        this.stream = textStream;
    }
    /**
     * Creates a processor instance asynchronously.
     *
     * @param ClassType - The constructor of the processor class
     * @param streamFunctions - Native functions for stream operations
     * @param statusStreamFunctions - Native functions for status operations
     * @param rest - Additional parameters for stream creation
     * @returns Promise resolving to a new processor instance
     * @throws InworldError if creation fails
     */
    static async createInstance(ClassType, streamFunctions, statusStreamFunctions, textStream, onDestroy = () => { }, ...rest) {
        let sentenceStreamInstance;
        let statusOrSentenceStream;
        try {
            statusOrSentenceStream = await streamFunctions.create(textStream, ...rest);
            if (!statusStreamFunctions.isOK(statusOrSentenceStream)) {
                throw new common_1.InworldError('Failed to create text stream: ' + ClassType.name, (0, helpers_1.getStatus)(statusOrSentenceStream));
            }
            const sentenceStream = statusStreamFunctions.getValue(statusOrSentenceStream);
            sentenceStreamInstance = new ClassType(sentenceStream, () => {
                streamFunctions.delete(sentenceStream);
                onDestroy();
            });
        }
        finally {
            statusStreamFunctions.delete(statusOrSentenceStream);
        }
        return sentenceStreamInstance;
    }
    /**
     * Returns the external reference to the native text stream implementation.
     *
     * @returns External reference object
     */
    getStream() {
        return new stream_1.TextStream(this.stream).getStream();
    }
    /**
     * Asynchronously reads the next item from the text stream.
     *
     * @returns Promise resolving to a result object with text content or done flag
     * @throws InworldError if reading from the stream fails
     */
    async next() {
        return new stream_1.TextStream(this.stream).next();
    }
    /**
     * Cleans up resources associated with this processor.
     */
    destroy() {
        if (this.stream) {
            __classPrivateFieldGet(this, _StreamTextProcessor_destroy, "f").call(this);
            this.stream = null;
            __classPrivateFieldSet(this, _StreamTextProcessor_destroy, null, "f");
        }
    }
}
exports.StreamTextProcessor = StreamTextProcessor;
_StreamTextProcessor_destroy = new WeakMap();
