// @inworld/runtime/build/graph/dsl/nodes/builtin/KeywordMatcherNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeywordMatcherNode = void 0;
const AbstractNode_1 = require("./AbstractNode");
/**
 * Keyword matcher node that matches keywords in text input.
 * This node processes text input and identifies matching keywords from a predefined list.
 *
 * @input {TextData} {@link TextData} - The data type that KeywordMatcherNode accepts as input
 * @output {MatchedKeywords} {@link MatchedKeywordsData} - The data type that KeywordMatcherNode accepts as output
 *
 * @example
 * ```typescript
 * const keywordNode = new KeywordMatcherNode({
 *   id: 'my-keyword-node',
 *   keywords: ['urgent', 'important', 'priority', 'asap'],
 *   reportToClient: true
 * });
 * ```
 */
class KeywordMatcherNode extends AbstractNode_1.AbstractNode {
    /**
     * Creates a new KeywordMatcherNode instance.
     *
     * @param props - Configuration for the keyword matcher node
     */
    constructor(props) {
        super(props);
        this.keywords = props.keywords;
    }
    /**
     * Converts the keyword matcher node to a graph configuration node.
     *
     * @returns The configured keyword matcher node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'KeywordMatcherNode',
            execution_config: {
                type: 'KeywordMatcherNodeExecutionConfig',
                properties: {
                    keywords: this.keywords,
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.KeywordMatcherNode = KeywordMatcherNode;
