// @inworld/runtime/build/graph/dsl/nodes/builtin/LLMChatRequestBuilderNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLMChatRequestBuilderNode = void 0;
const AbstractNode_1 = require("./AbstractNode");
/**
 * LLM chat request builder node that processes JSON input into formatted chat requests
 * using prompt templates. This node transforms structured JSON data into properly
 * formatted chat requests with messages, tools, and configuration.
 *
 * @input {JsonData} {@link JsonData} - The data type that LLMChatRequestBuilderNode accepts as input
 * @output {LLMChatRequest} {@link LLMChatRequestData} - The data type that LLMChatRequestBuilderNode accepts as output
 *
 * @example
 * ```typescript
 * const requestBuilderNode = new LLMChatRequestBuilderNode({
 *   id: 'chat-request-builder',
 *   messages: [
 *     { role: 'system', content: 'You are a helpful assistant.' },
 *     { role: 'user', content: '{{user_input}}' }
 *   ],
 *   tools: [
 *     { name: 'search', description: 'Search for information' }
 *   ],
 *   responseFormat: 'json'
 * });
 * ```
 */
class LLMChatRequestBuilderNode extends AbstractNode_1.AbstractNode {
    /**
     * Creates a new LLMChatRequestBuilderNode instance.
     *
     * @param props - Configuration for the LLM chat request builder node
     */
    constructor(props = {}) {
        super(props);
        this.messages = props.messages;
        this.tools = props.tools;
        this.toolChoice = props.toolChoice;
        this.responseFormat = props.responseFormat;
    }
    /**
     * Converts the LLM chat request builder node to a graph configuration node.
     *
     * @returns The configured LLM chat request builder node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'LLMChatRequestBuilderNode',
            execution_config: {
                type: 'LLMChatRequestBuilderNodeExecutionConfig',
                properties: {
                    messages: this.messages,
                    tools: this.tools,
                    tool_choice: this.toolChoice,
                    response_format: this.responseFormat,
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.LLMChatRequestBuilderNode = LLMChatRequestBuilderNode;
