// @inworld/runtime/build/common/data_types/api/tts_request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TTSRequest = void 0;
class TTSRequest {
    constructor(request) {
        this.request = request;
    }
    /**
     * Creates a TTSRequest from a TTSRequestData instance.
     *
     * @param data - The TTSRequestData instance
     * @returns A new TTSRequest instance
     */
    static from(data) {
        return new TTSRequest(data.getTTSRequest());
    }
    /**
     * Creates a TTSRequest with text content.
     *
     * @param text - The text content to synthesize
     * @param voice - Optional voice configuration
     * @returns A new TTSRequest instance
     */
    static withText(text, voice) {
        return new TTSRequest({
            contentType: 'text',
            text,
            voice,
        });
    }
    /**
     * Creates a TTSRequest with stream content.
     *
     * @param stream - The stream content to synthesize
     * @param voice - Optional voice configuration
     * @returns A new TTSRequest instance
     */
    static withStream(stream, voice) {
        return new TTSRequest({
            contentType: 'stream',
            stream,
            voice,
        });
    }
    /**
     * Checks if the request has text content.
     *
     * @returns True if the request contains text content
     */
    hasTextContent() {
        return this.request.contentType === 'text' && !!this.request.text;
    }
    /**
     * Checks if the request has stream content.
     *
     * @returns True if the request contains stream content
     */
    hasStreamContent() {
        return this.request.contentType === 'stream' && !!this.request.stream;
    }
    /**
     * Gets the text content if available.
     *
     * @returns The text content or undefined
     */
    getText() {
        return this.request.text;
    }
    /**
     * Gets the voice configuration if available.
     *
     * @returns The voice configuration or undefined
     */
    getVoice() {
        return this.request.voice;
    }
}
exports.TTSRequest = TTSRequest;
