// @inworld/runtime/build/common/data_types/type-mapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomOutputDataType = exports.CustomInputDataType = void 0;
exports.convertInputSpecsToTypes = convertInputSpecsToTypes;
exports.convertOutputSpecToType = convertOutputSpecToType;
exports.CustomInputDataTypeTyped = CustomInputDataTypeTyped;
exports.CustomOutputDataTypeTyped = CustomOutputDataTypeTyped;
/**
 * Enum representing the types of data that can be processed in the graph.
 */
var CustomInputDataType;
(function (CustomInputDataType) {
    CustomInputDataType["CUSTOM"] = "CUSTOM";
    CustomInputDataType["CLASSIFICATION_RESULT"] = "CLASSIFICATION_RESULT";
    CustomInputDataType["GOAL_ADVANCEMENT"] = "GOAL_ADVANCEMENT";
    CustomInputDataType["KNOWLEDGE_RECORDS"] = "KNOWLEDGE_RECORDS";
    CustomInputDataType["MEMORY_STATE"] = "MEMORY_STATE";
    CustomInputDataType["MATCHED_INTENTS"] = "MATCHED_INTENTS";
    CustomInputDataType["MATCHED_KEYWORDS"] = "MATCHED_KEYWORDS";
    CustomInputDataType["TEXT"] = "TEXT";
    CustomInputDataType["TEXT_STREAM"] = "TEXT_STREAM";
    CustomInputDataType["TTS_STREAM"] = "TTS_STREAM";
    CustomInputDataType["CHAT_REQUEST"] = "CHAT_REQUEST";
    CustomInputDataType["CONTENT"] = "CONTENT";
    CustomInputDataType["CONTENT_STREAM"] = "CONTENT_STREAM";
    CustomInputDataType["LIST_TOOLS"] = "LIST_TOOLS";
    CustomInputDataType["TOOL_CALLS_RESULTS"] = "TOOL_CALLS_RESULTS";
})(CustomInputDataType || (exports.CustomInputDataType = CustomInputDataType = {}));
/**
 * Enum representing the types of data that can be output from the graph.
 */
var CustomOutputDataType;
(function (CustomOutputDataType) {
    CustomOutputDataType["CUSTOM"] = "CUSTOM";
    CustomOutputDataType["TEXT"] = "TEXT";
    CustomOutputDataType["TTS"] = "TTS";
    CustomOutputDataType["CHAT_REQUEST"] = "CHAT_REQUEST";
    CustomOutputDataType["TEXT_STREAM"] = "TEXT_STREAM";
    CustomOutputDataType["SPEECH_CHUNK_STREAM"] = "SPEECH_CHUNK_STREAM";
    CustomOutputDataType["TTS_STREAM"] = "TTS_STREAM";
    CustomOutputDataType["LIST_TOOLS"] = "LIST_TOOLS";
    CustomOutputDataType["TOOL_CALLS_RESULTS"] = "TOOL_CALLS_RESULTS";
    CustomOutputDataType["TOOL_CALLS"] = "TOOL_CALLS";
})(CustomOutputDataType || (exports.CustomOutputDataType = CustomOutputDataType = {}));
/**
 * Helper function to convert InputTypeSpec array to CustomInputDataType array for runtime use.
 */
function convertInputSpecsToTypes(inputSpecs) {
    return inputSpecs.map((spec) => typeof spec === 'object' && '__type' in spec ? spec.__type : spec);
}
/**
 * Helper function to convert OutputTypeSpec to CustomOutputDataType for runtime use.
 */
function convertOutputSpecToType(outputSpec) {
    return typeof outputSpec === 'object' && '__type' in outputSpec
        ? outputSpec.__type
        : outputSpec;
}
/**
 * Helper function to create a typed custom input specification.
 * @template T - The custom TypeScript type for this input
 * @returns A CustomInput specification with the given type
 */
function CustomInputDataTypeTyped() {
    return {
        __type: CustomInputDataType.CUSTOM,
        __customType: undefined,
    };
}
/**
 * Helper function to create a typed custom output specification.
 * @template T - The custom TypeScript type for this output
 * @returns A CustomOutput specification with the given type
 */
function CustomOutputDataTypeTyped() {
    return {
        __type: CustomOutputDataType.CUSTOM,
        __customType: undefined,
    };
}
