// @inworld/runtime/build/graph/dsl/components/remote_stt_component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteSTTComponent = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_component_1 = require("./abstract_component");
/**
 * Remote STT (Speech-to-Text) component for audio transcription services.
 * Provides access to external speech recognition services to convert audio
 * input into text transcriptions.
 *
 * @example
 * ```typescript
 * const sttComponent = new RemoteSTTComponent({
\ *   sttConfig: {
 *     provider: 'whisper',
 *     modelName: 'whisper-1',
 *     language: 'en'
 *   }
 * });
 * ```
 */
class RemoteSTTComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new RemoteSTTComponent instance.
     *
     * @param {Object} props - Configuration for the remote STT component
     * @param {string} [props.id] - Optional explicit component identifier. Human-readable identifier would be generated if omitted.
     * @param {SpeechRecognitionConfig} props.sttConfig - STT configuration (camelCase required)
     */
    constructor(props) {
        super(props);
        this.sttConfig = props.sttConfig;
    }
    /**
     * Converts the remote STT component to a graph configuration component.
     *
     * @returns The configured remote STT component for graph execution
     */
    toGraphConfigComponent() {
        return {
            id: this.id,
            type: 'STTInterface',
            creation_config: {
                type: 'RemoteSTTConfig',
                properties: {
                    default_config: (0, snakify_1.default)(this.sttConfig),
                },
            },
        };
    }
}
exports.RemoteSTTComponent = RemoteSTTComponent;
