// @inworld/runtime/build/graph/dsl/components/RemoteTTSComponent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteTTSComponent = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const AbstractComponent_1 = require("./AbstractComponent");
/**
 * Remote TTS (Text-to-Speech) component for audio synthesis services.
 * This component provides access to external text-to-speech services
 * for converting text input into audio output with configurable voice parameters.
 *
 * @example
 * ```typescript
 * const ttsComponent = new RemoteTTSComponent({
 *   id: 'my-tts-component',
 *   synthesisConfig: {
 *     voice: 'alloy',
 *     speed: 1.0,
 *     pitch: 1.0
 *   }
 * });
 * ```
 */
class RemoteTTSComponent extends AbstractComponent_1.AbstractComponent {
    /**
     * Creates a new RemoteTTSComponent instance.
     *
     * @param props - Configuration for the remote TTS component
     */
    constructor(props) {
        super(props);
        this.synthesisConfig = props.synthesisConfig;
    }
    /**
     * Converts the remote TTS component to a graph configuration component.
     *
     * @returns The configured remote TTS component for graph execution
     */
    toGraphConfigComponent() {
        return {
            id: this.id,
            type: 'TTSInterface',
            creation_config: {
                type: 'RemoteTTSConfig',
                properties: {
                    synthesis_config: this.synthesisConfig
                        ? (0, snakify_1.default)(this.synthesisConfig)
                        : undefined,
                },
            },
        };
    }
}
exports.RemoteTTSComponent = RemoteTTSComponent;
