// @inworld/runtime/build/primitives/knowledge/remote_knowledge_config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteKnowledgeConfig = void 0;
const expose_binary_1 = require("../../expose_binary");
const knowledge_compile_config_1 = require("./knowledge_compile_config");
/**
 * Class for remote knowledge system configuration
 * @internal
 */
class RemoteKnowledgeConfig {
    /**
     * Creates a new RemoteKnowledgeConfig instance
     * @param config Configuration parameters
     */
    constructor(config) {
        this.knowledgeGetConfig = null;
        this.external = expose_binary_1.RemoteKnowledgeConfigFunctions.new();
        // Set API key
        expose_binary_1.RemoteKnowledgeConfigFunctions.setApiKey(this.external, config.apiKey);
        // Set language code if provided
        const configFunctions = expose_binary_1.RemoteKnowledgeConfigFunctions;
        if (config.languageCode &&
            typeof configFunctions.setLanguage === 'function') {
            configFunctions.setLanguage(this.external, config.languageCode);
        }
        // Create and set knowledge compile config
        this.knowledgeCompileConfig = new knowledge_compile_config_1.KnowledgeCompileConfig(config.knowledgeCompileConfig);
        expose_binary_1.RemoteKnowledgeConfigFunctions.setKnowledgeCompileConfig(this.external, this.knowledgeCompileConfig.getExternal());
        // Store knowledge get config as a plain object if provided
        if (config.knowledgeGetConfig) {
            this.knowledgeGetConfig = config.knowledgeGetConfig;
        }
    }
    /**
     * Gets the reference to the native object
     * @returns Native object reference
     */
    getExternal() {
        return this.external;
    }
    /**
     * Destroys the native object
     */
    destroy() {
        if (this.knowledgeCompileConfig) {
            this.knowledgeCompileConfig.destroy();
        }
        if (this.external) {
            expose_binary_1.RemoteKnowledgeConfigFunctions.delete(this.external);
            this.external = null;
        }
    }
}
exports.RemoteKnowledgeConfig = RemoteKnowledgeConfig;
